exercises/zebra-puzzle/.meta/src/reference/java/ZebraPuzzle.java

```diff
@@ -0,0 +1,108 @@
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+
+class ZebraPuzzle {
+    private String waterDrinker = "";
+    private String zebraOwner = "";
+
+    ZebraPuzzle() {
+        this.solve();
+    }
+
+    private static boolean isJustRightOf(int a, int b) {
+        return a - 1 == b;
+    }
+
+    private static boolean nextTo(int a, int b) {
+        return isJustRightOf(a, b) || isJustRightOf(b, a);
+    }
+
+    private void solve() {
```

Comment:
This method is a bit big. Would it be possible to extract some of it into individual methods? For example extract each method passed to `forEach`? 🙂

Reply:
The structure of this solution relies on lambda capture. Restructuring that into multiple methods seems impractical to me, since there would be 5*5 parameters for the last method.
The solution was heavily influenced by the solution provided in the Go and Python track, which was adapted from a solution by Peter Norvig.

Reply:
I don't think it'll be quite as bad as that 🙂 From a little playing around I think it looks like the last method will need at most 11 parameters. Not ideal but I think I might still prefer it to having one massive method. And it could be decreased by storing `houses` as a static variable and passing parameters as arrays, e.g. `int[] colorPermutation` instead of `int red, int green, int yellow...` if that makes sense? 🙂 

So I still think it would be a nice improvement. But you absolutely don't have to do it if you don't want to! 🙂 If you'd rather have it merged as is then that's fine, I might open an issue for it which can be picked up by anyone later 🙂 Does that sound okay? 🙂

Reply:
It looks great, thank you! 😄 

I see what you mean, jumping from method to method can make it harder to see the flow. However, I also think it's very hard to take in and understand a really long method. When it's broken up into smaller methods then you can try to understand each piece individually and then use that to understand what it all does together. I find that easier, but I guess it varies from person to person 🙂 

And as the solution to this exercise is a bit complicated (it being a logic puzzle) I guess no matter what we do the code will probably be a little bit difficult to understand 🙂